pkg/twwidth/cache.go

```diff
@@ -0,0 +1,254 @@
+package twwidth
+
+import (
+	"sync"
+	"sync/atomic"
+)
+
+const (
+	cacheMinCapacity = 1024
+	cacheMaxCapacity = 100000
+)
+
+type lruCacheKey string
+
+func (k lruCacheKey) String() string {
+	return string(k)
+}
+
+// cacheEntry represents a node in the doubly-linked LRU list.
+type cacheEntry struct {
+	key   lruCacheKey
+	value int
+	prev  *cacheEntry
+	next  *cacheEntry
+}
+
+// lruCache is a simple, bounded LRU cache for lruCacheKey -> int.
+//
+// It is fully thread-safe via a single mutex. For this use case
+// (string width caching), the mutex overhead is negligible compared
+// to Unicode/regex work.
+type lruCache struct {
+	mu       sync.Mutex
+	items    map[lruCacheKey]*cacheEntry
+	head     *cacheEntry // Most recently used
+	tail     *cacheEntry // Least recently used
+	capacity int
+	hits     int64 // atomic access for hit tracking
+	misses   int64 // atomic access for miss tracking
+}
+
+// newLRUCache creates a new LRU cache with the given capacity.
+// If capacity <= 0, the cache effectively disables itself.
+func newLRUCache(capacity int) *lruCache {
+	if capacity <= 0 {
```

Comment:
In what cases does someone want to disable caching? Should we return an error instead? Also, depending on the golang version, we can use built-in min and max functions

Reply:
@huangsam … on another note, I think I should not have `twwcache`, as there might be other sections ... such as `stringerCache map[reflect.Type]reflect.Value` ...  that may benefit from this cache.

Reply:
I see. What is your plan for caching across the whole app lifecycle? If it's long running, might need to be more careful about managing entries to avoid memory leaks